TWTValidation/JSON Validator/Parser/TWTJSONSchemaParser.m

```diff
@@ -0,0 +1,652 @@
+    node.minimumPropertyCount = [self parseUnsignedIntegerFromSchema:objectSchema forKey:kTWTJSONSchemaKeywordMinProperties];
+    NSArray *requiredNames = [self parseArrayOfUnqiueStringsFromSchema:objectSchema forKey:kTWTJSONSchemaKeywordRequired];
+    node.requiredPropertyNames = requiredNames ? [NSSet setWithArray:requiredNames] : nil;
+    node.propertySchemas = [self parseDictionaryOfSchemasFromSchema:objectSchema forKey:kTWTJSONSchemaKeywordProperties nodeClass:[TWTJSONSchemaNamedPropertyASTNode class]];
+    node.patternPropertySchemas = [self parseDictionaryOfSchemasFromSchema:objectSchema forKey:kTWTJSONSchemaKeywordPatternProperties nodeClass:[TWTJSONSchemaPatternPropertyASTNode class]];
+    node.additionalPropertiesNode = [self parseAdditionalItemsOrPropertiesWithSchema:objectSchema forKey:kTWTJSONSchemaKeywordAdditionalProperties];
+    node.propertyDependencies = [self parseDependenciesWithSchema:objectSchema];
+
+    return node;
+}
+
+
+#pragma mark - Keyword and value parser methods
+
+- (TWTJSONSchemaASTNode *)parseCommonKeywordsFromSchema:(NSDictionary *)schema toNode:(TWTJSONSchemaASTNode *)node
+{
+    node.schemaTitle = [self parseStringFromSchema:schema forKey:kTWTJSONSchemaKeywordTitle];
+    node.schemaDescription = [self parseStringFromSchema:schema forKey:kTWTJSONSchemaKeywordDescription];
+    NSArray *validValues = [self parseArrayFromSchema:schema forKey:kTWTJSONSchemaKeywordEnum];
+    node.validValues = validValues ? [NSSet setWithArray:validValues] : nil;
+    node.andSchemas = [self parseArrayOfSchemasFromSchema:schema forKey:kTWTJSONSchemaKeywordAllOf];
+    node.orSchemas = [self parseArrayOfSchemasFromSchema:schema forKey:kTWTJSONSchemaKeywordAnyOf];
+    node.exactlyOneOfSchemas = [self parseArrayOfSchemasFromSchema:schema forKey:kTWTJSONSchemaKeywordOneOf];
+    node.notSchema = [self parseSchemaFromSchema:schema forKey:kTWTJSONSchemaKeywordNot];
+    node.definitions = [self parseDictionaryFromSchema:schema forKey:kTWTJSONSchemaKeywordDefinitions];
+
+    return node;
+}
+
+// Keyword: type
+// Example: { "type" : "boolean" }
+//          { "type" : [ "string", "object", "null" ] }
+- (NSSet *)parseTypeKeywordFromSchema:(NSDictionary *)schema
+{
+    id typeValue = schema[kTWTJSONSchemaKeywordType];
+
+    if (!typeValue) {
+        return nil;
+    }
+
+    // Relies on parseSchema: having filtered out schemas with type array, object, string, number, or integer. Thus, type is either:
+    // A) One of these strings: "boolean", "null", "any"
+    // B) An array of valid type strings
+    // C) Invalid wrt the JSON Schema definition of the type keyword
+
+    [self pushPathComponent:kTWTJSONSchemaKeywordType];
+
+    static NSSet *allowedTypes = nil;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        allowedTypes = [NSSet setWithObjects:kTWTJSONSchemaTypeKeywordAny, kTWTJSONSchemaTypeKeywordBoolean, kTWTJSONSchemaTypeKeywordNull, nil];
+    });
```

Comment:
It’s generally good form to put your static variables at the top of your method. It reinforces the fact that the code is only executed the first time the method is invoked.